Emit inlining information to improve line numbers

ihnorton opened the pull request:
Also changes the parser to include file information with every linenode (see https://github.com/JuliaLang/julia/issues/1334#issuecomment-127648525)

before ([Tim's example](https://github.com/JuliaLang/julia/issues/1334#issuecomment-109832854)):

```
function foo()
    @show error("oops")
end

julia> foo3()
ERROR: oops
 in foo3 at /cmn/julia/base/userimg.jl:109
```

after:

```
ERROR: oops
 in foo2 at /cmn/julia/base/userimg.jl:4 
    [ from inlined function at ./show.jl:109 ]
```

[over 5 runs rebuilding sysimg](https://gist.github.com/ihnorton/2e0146d54d6aa0667e62) I get average times of:

master: 81.328s (max: 82.28)
branch: 81.846s (max: 82.76)

so, +.48s to rebuild sysimg (also +9MB max resident RAM)

ihnorton committed:
Include filename with every linenode

ihnorton committed:
Emit debug info for inlined code to improve line #s

Utilizes the DebugLoc "inlinedAt" parameter to remember inlining location in
addition to source code origin. When available, we now print both the inlinee
origin below the top-level call site.

Mostly addresses #1334. (this does not handle nested inlines, which may require
deeper changes)

yuyichao commented:
Would be better to construct a backtrace for inline functions as well but this is very nice to have.

Can the profiling system take advantage of this?

carnaval commented:
Didn't look in details but I'd be glad to have it as well. I don't think there will be much conflict but can this be done on top of the codegen rewrite ? Even if it should not really touch the same parts I think it's good in general. (Might also help in getting it merged ;-))

timholy commented:
I'm honored that you thought my `foo` function was worthy of being built into your local julia :smile:.

Seriously, this is really nice---a huge gain in terms of usability.

jakebolewski commented:
:cake:

ScottPJones commented:
👏

ihnorton commented:
@timholy I forgot to mention, this works with `precompile`:
```
isaiah@titan:/cmn/julia/base$ nm ../usr/lib/julia/sys.so | grep foo2
00000000002eac80 t jlcall_foo2_20918
00000000002eab50 t julia_foo2_20918
```

@carnaval it should be trivial to rebase on the codegen rewrite if needed, there's only about ~10 loc change in codegen.

tkelman commented on src/codegen.cpp:
```diff
@@ -4502,6 +4502,7 @@ static Function *emit_function(jl_lambda_info_t *lam)
     int prevlno = -1;
     for(i=0; i < stmtslen; i++) {
         jl_value_t *stmt = jl_cellref(stmts,i);
+        /*
```
should this block just be deleted, or does it need to be brought back?

ihnorton replied:
Forgot about this. If Jeff approves the parser change then all LineNode-handling code can be deleted.

jakebolewski replied:
I don't see what is controversial about the LineNode change.

vtjnash commented:
since this touches only the debuginfo parts of codegen, i expect it will merge cleanly with my codegen restructuring PR.

carnaval commented:
@vtjnash shhh don't tell, I thought if we pile up PRs depending on this one it would get merged faster

ihnorton commented:
Test failures are fixed.

tkelman commented:
Sweet! Will be good to squash out the failing commits.

ihnorton commented:
Squashed, rebased, and LineNumberNode excised. Unless someone (...@mlubin) wants to fight to get this in 0.4, I'll just let it bide until we branch.

jakebolewski commented:
You have a whitespace error.  Why wouldn't we merge this? 0.4 will probably be at least 8-12 months, so it seems worth it to have better debug info during that time.

mlubin commented:
Well I (and many other people) would very much appreciate backtraces working with macros in 0.4, but I'm not in a position to judge how risky it is to merge this.
Are there any unit tests we can add for these backtraces?

IainNZ commented:
Anything to make line numbers better is worth it

timholy commented:
Since we seem to be taking longer to branch than I expected, then heck, if it's all working there's no reason to have this languish (I'm not really qualified to review it). I recognize touching codegen is risky, but this is a pretty big usability problem (I had someone in my lab ask about this just today).

StefanKarpinski commented:
I'm in favor of merging if @ihorton thinks it's safe.

ihnorton commented:
- updated based on discussion with @vtjnash and @jakebolewski: now adds `file` field to `LineNumberNode`, and uses `LineNumberNode` everywhere
- now [handle inlines from the same file](https://github.com/JuliaLang/julia/commit/e2a8d729585bc23676a3835d49b057fce02dba31#diff-6d4d21428a67320600faf5a1a9f3a16aR4595)
- added test for inlining information

I am comfortable merging this, but would appreciate a once-over/thumbs-up from anyone who can review (especially the changes to ast.c and codegen.cpp).

vtjnash commented on test/backtrace.jl:
```diff
@@ -5,10 +5,29 @@ have_backtrace = false
+
+    try
+        testinline()
+    catch err
```
> LoadError("C:\\projects\\julia\\test\\backtrace.jl",18,ErrorException("invalid redefinition of constant err"))

looks like CI wants you to pick a different name here (or add err to the let block variable list)

ihnorton replied:
must be an ordering thing because it did pass on several workers. I just added `err` to the `let` block, and also mimicked the handling in replutil (skip `fromC` and `:error` frames) because I don't think `bt[4]` is accurate for all platforms (namely win32).

vtjnash commented on base/docs/Docs.jl:
```diff
@@ -287,7 +287,7 @@ isexpr(x, ts...) = any(T->isa(T, Type) && isa(x, T), ts)
 function unblock(ex)
     isexpr(ex, :block) || return ex
-    exs = filter(ex->!isexpr(ex, :line), ex.args)
+    exs = filter(ex -> !(isa(ex, LineNumberNode) || isexpr(ex, :line)), ex.args)
```
is the `isexpr(:line)`test still needed?

ihnorton replied:
yes, in macro expansions we call `scm_to_julia_` with `expronly=1`, so `:line` (and all other tree nodes) get translated to Exprs.

vtjnash commented on src/ast.c:
```diff
@@ -460,7 +462,10 @@ static value_t julia_to_scm_(jl_value_t *v)
-        return julia_to_list2((jl_value_t*)line_sym, jl_fieldref(v,0));
+        return fl_cons(julia_to_scm_((jl_value_t*)line_sym),
+                       fl_cons(julia_to_scm_(jl_fieldref(v,1)),
+                       fl_cons(julia_to_scm_(jl_fieldref(v,0)), FL_NIL))
```
i think you are missing flisp gc-roots for the intermediate `julia_to_scm_` and `fl_cons` calls (remember, C does not guarantee right-to-left execution within a statement so you need https://github.com/JuliaLang/julia/pull/12544/files#diff-ee48e36e4d097c33660f778b57339696R416)

ihnorton replied:
updated to root the sym and use `julia_to_list2` for the args.

vtjnash commented on src/codegen.cpp:
```diff
@@ -4055,32 +4055,17 @@ static Function *emit_function(jl_lambda_info_t *lam)
     ctx.dbuilder = &dbuilder;
 #ifdef LLVM37
-    DIFile *fil = NULL;
+    DIFile *topfil = NULL;
```
style nit: i would prefer `topfile` here so you don't have to wonder if and how you abbreviated the name (and the reader doesn't have to wonder what FIL means here). plus vowels are free.

ihnorton replied:
lol. done.

vtjnash commented on src/codegen.cpp:
```diff
@@ -4578,59 +4562,59 @@ static Function *emit_function(jl_lambda_info_t *lam)
+                if ((dfil == topfil || dfil == NULL) &&
+                    lno >= ctx.lineno) // if we are in the top-level file
+                                       // and the current lineno is less than
+                                       // the last, this is same-file inline
```
i think we can improve this heuristic as a todo item, but this should be good enough for now

ihnorton replied:
added a TODO note. I also though about indicating entry/exit from inlining with negative line numbers but I don't understand the lowering passes well enough to do that myself yet.

vtjnash commented on src/debuginfo.cpp:
```diff
@@ -491,6 +499,13 @@ void lookup_pointer(DIContext *context, char **name, size_t *line,
+
+    if (inlineinfo.getNumberOfFrames() > 1) {
+        topinfo = inlineinfo.getFrame(inlineinfo.getNumberOfFrames() - 1);
+        if (strcmp(topinfo.getFunctionName(), "<invalid>") == 0) goto done;
```
jumping to done will delete this frame. i think it might be better to just allow "<invalid>" to show up if this is where the code actually managed to arrive at

vtjnash commented on src/disasm.cpp:
```diff
@@ -228,10 +228,12 @@ int OpInfoLookup(void *DisInfo, uint64_t PC,
     int fromC;
-    jl_getFunctionInfo(&name, &line, &filename, pointer, &fromC, skipC);
+    jl_getFunctionInfo(&name, &filename, &line, &inlined_file, &inlined_line, pointer, &fromC, skipC, 1);
     free(filename);
```
add `free(inlined_file)` here also

vtjnash commented on src/debuginfo.cpp:
```diff
@@ -861,10 +880,22 @@ void jl_getFunctionInfo(char **name, size_t *line, char **filename,
         }
+
+        DILexicalBlockFile locscope = DILexicalBlockFile(prev.Loc.getScope((*it).second.func->getContext()));
+        jl_copy_str(filename, (char*)(uintptr_t)locscope.getFilename().data());
```
use `.c_str()` to ensure the string is null-terminated (instead of `.data()`)

vtjnash replied:
also, don't add unnecessary casts, those can hide bugs (`(char*)(uintptr_t)`)

ihnorton replied:
```
debuginfo.cpp:884:72: error: ‘class llvm::StringRef’ has no member named ‘c_str’
         jl_copy_str(filename, (char*)(uintptr_t)locscope.getFilename().c_str());
```

(related discussion: https://github.com/JuliaLang/julia/pull/12433#discussion_r36047863)

ihnorton replied:
removed the cast (not sure what I was thinking there)

vtjnash replied:
> Agreed. We seem to have gotten away with it so far.

Interesting thought, but we've also had a number of over-reads on strings (sometimes turning into segfaults) from the usage of `.data()` instead of `.c_str()`. For StringRef, use `.str().c_str()`, or just modify `jl_copy_str` to take the length as an argument since the information should have always been present anyways.

ihnorton replied:
> use .str().c_str()

ok, done.

yuyichao replied:
Is this valid? IIUC, `llvm::StringRef::str()` returns a [new `std::string` object](http://llvm.org/docs/doxygen/html/classllvm_1_1StringRef.html#ab02cb6f4884b5c788efb97b6741dccaf) which should get out of scope and be destructed after `c_str()` returns.

I've never dig into the livetime of c++ temporaries so I might be wrong here.

ihnorton replied:
Seems ok: http://stackoverflow.com/questions/2506793/c-life-span-of-temporary-arguments

also LLVM does it:
```
isaiah@titan:/cmn/jldev/deps/llvm-svn/lib$ grep -r "str().c_str" *
Analysis/Analysis.cpp:    *OutMessages = strdup(MsgsOS.str().c_str());
ExecutionEngine/ExecutionEngineBindings.cpp:                                       SectionName.str().c_str());
ExecutionEngine/ExecutionEngineBindings.cpp:                                       SectionName.str().c_str(),
LineEditor/LineEditor.cpp:  Data->EL = ::el_init(ProgName.str().c_str(), In, Out, Err);
LTO/LTOCodeGenerator.cpp:    CodegenOptions.push_back(strdup(o.first.str().c_str()));
Support/Unix/Path.inc:      ssize_t len = readlink(aPath.str().c_str(), exe_path, sizeof(exe_path));
Support/Unix/Program.inc:    int Err = posix_spawn(&PID, Program.str().c_str(), FileActions,
Support/Unix/Signals.inc:        OS << format("%s ", FunctionName.str().c_str());
Support/Unix/Signals.inc:        OS << format("%s", FileLineInfo.str().c_str());
Target/NVPTX/NVPTXReplaceImageHandles.cpp:    Idx = MFI->getImageHandleSymbolIndex(NewSymStr.str().c_str());
Target/NVPTX/NVPTXRegisterInfo.h:    return getStrPool()->getManagedString(O.str().c_str())->c_str();
Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:        llvm_unreachable(RSO.str().c_str());
```

vtjnash commented on src/gf.c:
```diff
@@ -916,10 +916,11 @@ DLLEXPORT jl_function_t *jl_instantiate_staged(jl_methlist_t *m, jl_tupletype_t
-    jl_expr_t *linenode = jl_exprn(line_sym, 2);
+    jl_value_t *linenode = jl_new_struct(jl_linenumbernode_type,
+                                         m->func->linfo->file,
+                                         jl_box_long(m->func->linfo->line)
```
this box needs a gc-root before the call to `jl_new_struct`

vtjnash commented on src/jltypes.c:
```diff
@@ -3345,8 +3345,8 @@ void jl_init_types(void)
-                        jl_svec(1, jl_symbol("line")),
-                        jl_svec(1, jl_long_type), 0, 0, 1);
+                        jl_svec(2, jl_symbol("file"), jl_symbol("line")),
+                        jl_svec(2, jl_symbol_type, jl_long_type), 0, 0, 1);
```
that last "1" can be updated to "2" (both fields are always initialized)

vtjnash commented on src/task.c:
```diff
@@ -737,7 +745,10 @@ DLLEXPORT void gdblookup(ptrint_t ip)
+    size_t inlined_line;
+    char *inlined_file;
+    frame_info_from_ip(&func_name, &file_name, &line_num, &inlined_file, &inlined_line, ip,
+                      /* skipC */ 0, /* skipInline */ 1);
```
it looks like we need a `free` call for `file_name` and `inlined_file` at the end of this function

vtjnash commented on test/backtrace.jl:
```diff
@@ -5,10 +5,29 @@ have_backtrace = false
     end
 end
 
-@test have_backtrace
```
what happened to this test? might be good to add a test for the other properties in the backtrace to show that it found the right point in the file and didn't think it had inlining present.

ihnorton replied:
accidentally removed. I added some more sophisticated tests.

vtjnash commented on test/backtrace.jl:
```diff
@@ -5,10 +5,29 @@ have_backtrace = false
+
+let inlinedfile, inlinedline
+    function testinline()
+        @issue1334_inlinedcode
```
i might suggest adding a few lines to this function to make the backtrace test a little harder. for example:
```julia
function
  1
  @issue1334
  2
end
```

ihnorton replied:
I went a slightly different direction with this for now, generating the whole function expression with fixed/known line numbers.

(https://github.com/JuliaLang/julia/issues/8066 would be really helpful here)

ihnorton commented:
Alright, I think I addressed everything. @vtjnash thanks for the thorough review. Hopefully it passes now on Windows.

mlubin commented:
:+1:

vtjnash commented on src/ast.c:
```diff
@@ -459,8 +461,14 @@ static value_t julia_to_scm_(jl_value_t *v)
-    if (jl_typeis(v, jl_linenumbernode_type))
-        return julia_to_list2((jl_value_t*)line_sym, jl_fieldref(v,0));
+    if (jl_typeis(v, jl_linenumbernode_type)) {
+        value_t args = julia_to_list2(jl_fieldref(v,1), jl_fieldref(v,0));
```
i think you are still missing a gc-root for `args` during the call to `julia_to_scm_`

ihnorton replied:
fixed

vtjnash commented on src/ast.c:
```diff
@@ -326,9 +326,11 @@ static jl_value_t *scm_to_julia_(value_t e, int eo)
                     return jl_new_struct(jl_linenumbernode_type,
-                                         scm_to_julia_(car_(e),0));
+                                         scm_to_julia_(car_(cdr_(e)),0), // filename
+                                         scm_to_julia_(car_(e),0)        // line
```
this line number box is also missing a gc-root during the call to jl_new_struct (looks like it was missing before too :frowning:)

yuyichao replied:
Hmmm I guess it didn't cause an issue before (and escaped the gc stress test) because it hits the pre-boxed cache most of the time?

ihnorton replied:
so are the `jl_new_struct` calls below missing boxes too?

vtjnash commented on src/debuginfo.cpp:
```diff
@@ -491,6 +499,12 @@ void lookup_pointer(DIContext *context, char **name, size_t *line,
+        topinfo = inlineinfo.getFrame(inlineinfo.getNumberOfFrames() - 1);
+        jl_copy_str(inlined_file, topinfo.getFileName());
+        *inlined_line = topinfo.getLine();
+    }
```
it looks like this block needs to be cloned to the LLVM35+ block below.

ihnorton replied:
moved out of the ifdef entirely, because there should be nothing version-specific here.

vtjnash replied:
that would be nice, but i thought it was `info.Filename.c_str()` on the new versions?

vtjnash commented on src/ast.c:
```diff
@@ -326,9 +326,14 @@ static jl_value_t *scm_to_julia_(value_t e, int eo)
+                if (sym == line_sym && n==2) {
+                    jl_value_t *filename = scm_to_julia_(car_(cdr_(e)),0);
+                    jl_value_t *linenum  = scm_to_julia_(car_(e),0);
+                    JL_GC_PUSH2(&filename, &linenum);
```
note that assuming filename is a Symbol, it doesn't need a root. If it wasn't a symbol, this root would have been needed before the call to create the box to linenum.

vtjnash commented on src/ast.c:
```diff
@@ -326,9 +326,14 @@ static jl_value_t *scm_to_julia_(value_t e, int eo)
+                    return temp;
                 }
                 if (sym == label_sym) {
                     return jl_new_struct(jl_labelnode_type,
```
:person_frowning: yes all of these need are missing a gc root for their argument

ihnorton committed:
Add missing GC roots in ast.c

ihnorton commented:
Updated: tested/fixed on LLVM37, and added missing roots.

ihnorton commented:
(cc @Keno since you seemed to be around at the moment, and this is your wheelhouse)

timholy commented:
Bump. @Keno, I notice you haven't reviewed many patches lately, and this would be quite a stellar one to help out with. If this could be merged for 0.4, it would improve the quality of life for many people.

Keno commented:
I'm all on board, but @JeffBezanson needs to approve the size increase in the AST.

timholy commented:
Ah, great, many thanks!!

IainNZ commented:
Does this help the line number issues with macros, or is this just focussed on inlined functions?

ihnorton commented:
This PR will give you the code-origin information for most macros; I think there are some instances where macros defined with the short quote form `:(   )` do not have any `:line` information attached, but that should be fixable. Nor will manually-constructed `Expr`s unless you attach the line info  yourself (see the tests in #12727). There are two bigger limitations right now:

- No intermediate information for nested macros. That is, if `function foo()` calls `@macro1()` which in turn calls `@macro2()`, you get:
  ```
  ERROR: oops
   in foo at myfile.jl:100
   [ thrown by inlined function defined at ]
   /wherever/macro2/is/defined.jl:macro2_line_number
  ```
  There will be no information in the trace for `macro1`. Accurately identifying the call-site is already a big win IMO -- [see an example from a failed appveyor build](https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.8203/job/t9f83y8xc3t1sd2s#L1061). (tracing arbitrary nesting may require some deeper changes)

- inlined functions (`@inline`) will not have code-origin line information, because inference [strips :line expressions](https://github.com/JuliaLang/julia/blob/f0c823922b3f78dd95a3e5262f3fa1139c3571b9/base/inference.jl#L2262-L2263) (cc @vtjnash, would be nice to change that. I tried setting `body.args = ast.args[3].args::Array{Any,1}` but inference croaks with a nice 10k line dump)

(for reference, [here is a longer overview of the system](https://gist.github.com/ihnorton/11efc50b6ff0931d2b68), and the origin of the issue in #1334)

vtjnash commented:
what's at the top of that? If I had to guess, I would say
https://github.com/JuliaLang/julia/blob/f0c823922b3f78dd95a3e5262f3fa1139c3571b9/base/inference.jl#L2572-L2579 might get confused if there is a LineNumberNode at the end of a function

ihnorton commented:
> what's at the top of that 10 line dump? 

not sure, it's actually > 5k lines, because that was the size of the term buffer. I'll check in gdb later.  does the inliner reorder expressions? if so then that makes sense.

vtjnash commented:
no, it shouldn't be reordering them. the frames that matter should be right at the top of that terminal window ... i wish gterm used a higher default.

vtjnash commented:
this looks ready to merge to me. it looks like someone will need to deal with some merge conflict on the command line though to push it to master.

prcastro commented:
:+1:

JeffBezanson commented on base/cartesian.jl:
```diff
@@ -222,7 +222,7 @@ function poplinenum(ex::Expr)
         if length(ex.args) == 1
             return ex.args[1]
-        elseif length(ex.args) == 2 && ex.args[1].head == :line
+        elseif length(ex.args) == 2 && isa(ex.args[1], LineNumberNode)
```
Doesn't this need to work on surfa

ihnorton replied:
Fixed (I think). Thanks.

JeffBezanson commented on base/deprecated.jl:
```diff
@@ -80,7 +80,7 @@ function firstcaller(bt::Array{Ptr{Void},1}, funcsym::Symbol)
             continue
         end
-        fname, file, line, fromC = lkup
+        fname, file, line, inlined_file, inlined_line, fromC = lkup
```
What do the `inlined_file` and `inlined_line` fields mean? It seems potentially annoying to deal with extra information about inlining for each source location. I would expect inlined frames to be represented the same as normal ones, perhaps with an `inline` flag set to `true` or something.

ihnorton replied:
They mean "where the code was inlined from".

> It seems potentially annoying to deal with extra information about inlining for each source location. I would expect inlined frames to be represented the same as normal ones, perhaps with an inline flag set to true or something.

Each address in the backtrace may have inlining information associated with it; there is no separate frame ip for inlined code. Returning the inlining info here seemed simplest option to fit in the current implementation of `jl_getFunctionInfo`/`lookup_code_address`.

`jl_getFunctionInfo` may eventually need a redesign to return a chain of inline information (for nested code), but this PR is not emitting that (intermediates) debug info either.

JeffBezanson replied:
Yes, there is no separate frame ip, but there is a separate source location record (conceptually speaking).

I still don't quite understand this. The printing code shows `inlinedfile : inlinedline`, *then* says `by code inlined from: ` and shows `file : line`. That seems to contradict inlinedfile/inlinedline meaning where the code was inlined from. Or I might be misinterpreting something.

ihnorton replied:
You are right, the variable names were not good. I renamed them.

- `file` and `line` always reflect where code at a given address was defined, inlined or otherwise.
- `inlinedat_file` and `inlinedat_line` (renamed) reflect where that code ended up.

With updates, here is an example backtrace:
```
julia> foo()
ERROR: foobar
 [inlined code] from show.jl:110
 in foo at /cmn/jlbugs/1334_macro_lnos/foo.jl:5
```

for the following function:
```
function foo()
    @show error("foobar")
end
```

ihnorton replied:
> there is a separate source location record (conceptually speaking).

Does the trace above address this sufficiently? (thanks @yuyichao)

JeffBezanson commented:
What's the impact on the size of the system image?

JeffBezanson commented on base/replutil.jl:
```diff
@@ -328,9 +328,19 @@ function show_method_candidates(io::IO, ex::MethodError)
+
+    if (inlinedfile != symbol(""))
+        print(io, " at ", inlinedfile, ":", inlinedline)
+        print(io, " \n    [ thrown by inlined function defined at ]\n    ")
```
I find this output a bit confusing. Since the brackets close before all the info is printed, it looks like the location is simply missing. I would print it something like `in inlined function defined at file:line`.

ihnorton replied:
how about `by code inlined from`? could be from macro or inline.

JeffBezanson replied:
Yes, that makes sense.

ihnorton committed:
Address review comments

- renamed inlinedfile/line -> inlinedat_file/line
- update inlining message
- fix poplinenum.

ihnorton commented:
> What's the impact on the size of the system image?

~ 700k

```
master:
isaiah@titan:/cmn/julia$ ls -la usr/lib/julia/*
-rw-r--r-- 1 isaiah isaiah  1892596 Aug 27 00:34 usr/lib/julia/inference0.ji
-rw-r--r-- 1 isaiah isaiah  1832530 Aug 27 00:35 usr/lib/julia/inference.ji
-rw-rw-r-- 1 isaiah isaiah 30183440 Aug 27 00:36 usr/lib/julia/sys.o
-rwxrwxr-x 1 isaiah isaiah 25856974 Aug 27 00:36 usr/lib/julia/sys.so

this branch:
isaiah@titan:/cmn/julia$ ls -la usr/lib/julia/*
-rw-r--r-- 1 isaiah isaiah  1902264 Aug 27 00:01 usr/lib/julia/inference0.ji
-rw-r--r-- 1 isaiah isaiah  1848958 Aug 27 00:02 usr/lib/julia/inference.ji
-rw-rw-r-- 1 isaiah isaiah 31107296 Aug 27 00:15 usr/lib/julia/sys.o
-rwxrwxr-x 1 isaiah isaiah 26560922 Aug 27 00:15 usr/lib/julia/sys.so
```

ihnorton commented:
Rebased.

JeffBezanson commented on base/profile.jl:
```diff
@@ -134,8 +136,8 @@ maxlen_data() = convert(Int, ccall(:jl_profile_maxlen_data, Csize_t, ()))
 
 function lookup(ip::Ptr{Void})
```
We should probably make more use of this interface to `jl_lookup_code_address`.

JeffBezanson commented:
Hmm, I guess 2-3% isn't so bad. The `MODE_AST` serializer needs to do a better job compressing symbols.

carnaval commented:
(quick&dirty test: lz4 shrinks sys.so by 4x and only takes ~50ms to decompress, so it might be easier than spending too much time improving the serializer)

dcjones commented:
Nothing of substance to add, but I found this branch to be exceedingly useful to profile some macro heavy code that previously reported complete nonsense. I'm unreasonably excited to see it merged.

timholy commented:
Woot! Didn't expect this to make it in. But really looking forward to it (building now).

ScottPJones commented:
:+1: :100: :clap: :party:
